README.md
# Flutter Screen Theme plugin

This plugin for [Flutter](https://flutter.io) adds the ability to change the status bar theme: light or dark (Android and iOS).
It also provides the ability to change the navigation bar theme and color (Android only)

<p align="center">
  <img src="https://raw.githubusercontent.com/g123k/flutter-screen-theme-plugin/master/assets/demo.gif" alt="Demo App" style="margin:auto" width="540" height="960">
</p>


## Getting Started

On iOS only, please add in your Info.plist:
```xml
<key>UIViewControllerBasedStatusBarAppearance</key>
<false/>
```

Then you just have to import the package in your dart files with
```dart
import 'package:screentheme/screentheme.dart';
```

For each call, you can specify which platform you want to target (by default both):
```dart
ScreenTheme.darkStatusBar(platform: Platform.Android);
```

Each call returns a boolean, which is `true` when the platform and the OS version support the feature.
For the navigation bar, all calls will throw a `MissingPluginException`.

## Status bar

**Compatibility**: Android (6.0+) & iOS

On Android, it will only work with Android 6.0 (Marshmallow) and above devices.


## Navigation bar

**Compatibility**: Android only

Android 5.0 (Lollipop) and above: color

Android 8.0 (Oreo) and above: theme (dark/light)


CHANGELOG.md
## [0.0.2] - 21th March 2018.

* Fix a bug when using Dart 2.


## [0.0.1] - 18th March 2018.

* Initial Release.


example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:screentheme_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new MyApp());

    // Verify that platform version is retrieved.
    expect(
        find.byWidgetPredicate(
          (Widget widget) =>
              widget is Text && widget.data.startsWith('Running on:'),
        ),
        findsOneWidget);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:screentheme/screentheme.dart';

void main() => runApp(new MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => new _MyAppState();
}

class _MyAppState extends State<MyApp> {
  bool _lightStatusBar = false;
  bool _lightNavigationBar = false;

  @override
  initState() {
    super.initState();
    _init();
  }

  _init() async {
    await _initStatusBar();
    await _initNavigationBar();
  }

  _initStatusBar() async {
    bool lightStatusBar;
    try {
      lightStatusBar = await ScreenTheme.hasALightStatusBar();
    } on Exception {
      lightStatusBar = false;
    }

    setState(() {
      _lightStatusBar = lightStatusBar;
    });
  }

  _initNavigationBar() async {
    bool lightNavigationBar;
    try {
      lightNavigationBar = await ScreenTheme.hasALightNavigationBar();
    } on Exception {
      lightNavigationBar = false;
    }

    setState(() {
      _lightNavigationBar = lightNavigationBar;
    });
  }

  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(
          title: new Text('Screen theme plugin example'),
        ),
        body: new Center(
          child: new Column(
            mainAxisSize: MainAxisSize.max,
            children: <Widget>[
              new Expanded(
                child: new Column(
                  mainAxisSize: MainAxisSize.max,
                  mainAxisAlignment: MainAxisAlignment.center,
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: <Widget>[
                    new Padding(
                      padding: const EdgeInsets.only(bottom: 20.0),
                      child: new Text("Light status bar: $_lightStatusBar"),
                    ),
                    new RaisedButton(
                        onPressed: () {
                          enableLightStatusBar();
                        },
                        child: new Text("Light theme")),
                    new RaisedButton(
                      onPressed: () {
                        enableDarkStatusBar();
                      },
                      child: new Text("Dark theme"),
                    )
                  ],
                ),
              ),
              new Expanded(
                child: new Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: <Widget>[
                    new Padding(
                      padding: const EdgeInsets.only(bottom: 20.0),
                      child: new Text(
                          "Light navigation bar: $_lightNavigationBar"),
                    ),
                    new RaisedButton(
                        onPressed: () {
                          enableLightNavigationBar();
                        },
                        child: new Text("Light theme")),
                    new RaisedButton(
                      onPressed: () {
                        enableDarkNavigationBar();
                      },
                      child: new Text("Dark theme"),
                    ),
                    new RaisedButton(
                      onPressed: () {
                        changeNavigationBarColor();
                      },
                      child: new Text("Red color"),
                    )
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  changeNavigationBarColor() async {
    await ScreenTheme.updateNavigationBarColor(Colors.red);
  }

  enableLightStatusBar() async {
    await ScreenTheme.lightStatusBar();
    bool hasLightStatusBar = await ScreenTheme.hasALightStatusBar();

    setState(() {
      _lightStatusBar = hasLightStatusBar;
    });
  }

  enableDarkStatusBar() async {
    await ScreenTheme.darkStatusBar();
    bool hasLightStatusBar = await ScreenTheme.hasALightStatusBar();

    setState(() {
      _lightStatusBar = hasLightStatusBar;
    });
  }

  enableLightNavigationBar() async {
    await ScreenTheme.lightNavigationBar();
    bool hasLightNavigationBar = await ScreenTheme.hasALightNavigationBar();

    setState(() {
      _lightNavigationBar = hasLightNavigationBar;
    });
  }

  enableDarkNavigationBar() async {
    await ScreenTheme.darkNavigationBar();
    bool hasLightNavigationBar = await ScreenTheme.hasALightNavigationBar();

    setState(() {
      _lightNavigationBar = hasLightNavigationBar;
    });
  }
}


ios/Classes/ScreenThemePlugin.m
#import "ScreenThemePlugin.h"

@implementation ScreenThemePlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"g123k/screentheme"
            binaryMessenger:[registrar messenger]];
  ScreenThemePlugin* instance = [[ScreenThemePlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary *args = call.arguments;
    NSString *platform = [args objectForKey:@"platform"];

    if (!([@"ios" isEqualToString:platform] || [@"both" isEqualToString:platform])) {
        result(FlutterMethodNotImplemented);
        return;
    }

    if ([@"changeStatusBarTheme" isEqualToString:call.method]) {
        NSString *theme = [args objectForKey:@"theme"];

        if ([@"light" isEqualToString:theme]) {
            [UIApplication sharedApplication].statusBarStyle = UIStatusBarStyleLightContent;
            result(@YES);
        } else if ([@"dark" isEqualToString:theme]) {
            [UIApplication sharedApplication].statusBarStyle = UIStatusBarStyleDefault;
            result(@YES);
        } else {
            result(@NO);
        }
    } else if ([@"hasLightStatusBar" isEqualToString:call.method]) {
        if ([UIApplication sharedApplication].statusBarStyle != UIStatusBarStyleDefault) {
            result(@YES);
        } else {
            result(@NO);
        }
    } else {
        result(FlutterMethodNotImplemented);
    }
}

@end


android/src/main/java/fr/g123k/screentheme/ScreenThemePlugin.java
package fr.g123k.screentheme;

import android.app.Activity;
import android.content.res.Resources;
import android.graphics.Color;
import android.os.Build;
import android.util.Log;
import android.view.View;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/**
 * Screen theme (status bar & navigation bar) Plugin
 */
public class ScreenThemePlugin implements MethodCallHandler {

    private static final String METHOD_CHANNEL_NAME = "g123k/screentheme";

    private final Activity activity;

    public ScreenThemePlugin(Activity activity) {
        this.activity = activity;
    }

    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), METHOD_CHANNEL_NAME);
        channel.setMethodCallHandler(new ScreenThemePlugin(registrar.activity()));
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        if (call.hasArgument("platform")) {
            String platform = call.argument("platform");
            if (!("android".equalsIgnoreCase(platform) || "both".equalsIgnoreCase(platform))) {
                result.error("PLATFORM", "This call is not for an Android device!", null);
                return;
            }
        }

        if (call.method.equals("changeStatusBarTheme") && call.hasArgument("theme")) {
            String theme = call.argument("theme");
            if ("light".equals(theme)) {
                result.success(lightStatusBar());
            } else if ("dark".equals(theme)) {
                result.success(darkStatusBar());
            } else {
                result.notImplemented();
            }
        } else if (call.method.equals("hasLightStatusBar")) {
            result.success(hasLightStatusBar());
        } else if (call.method.equals("changeNavigationBarTheme") && call.hasArgument("theme")) {
            String theme = call.argument("theme");
            if ("light".equals(theme)) {
                result.success(lightNavigationBar());
            } else if ("dark".equals(theme)) {
                result.success(darkNavigationBar());
            } else {
                result.notImplemented();
            }
        } else if (call.method.equals("hasLightNavigationBar")) {
            result.success(hasLightNavigationBar());
        } else if (call.method.equals("changeNavigationBarColor") && call.hasArgument("red") && call.hasArgument("green") && call.hasArgument("blue")) {
            Object red = call.argument("red");
            Object green = call.argument("green");
            Object blue = call.argument("blue");
            result.success(changeNavigationBarColor(Integer.parseInt(red.toString()), Integer.parseInt(green.toString()), Integer.parseInt(blue.toString())));
        } else if (call.method.equals("hasNavigationBar")) {
            result.success(hasANavigationBar());
        } else {
            result.notImplemented();
        }
    }

    private boolean supportsStatusBarTheming() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M;
    }

    private boolean hasLightStatusBar() {
        if (supportsStatusBarTheming()) {
            final View view = getView();
            int flags = view.getSystemUiVisibility();
            return (flags & View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR) != View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
        }
        return true;
    }

    // White status bar = default status bar theme for Android
    private boolean lightStatusBar() {
        if (supportsStatusBarTheming()) {
            final View view = getView();
            int flags = view.getSystemUiVisibility();
            flags &= ~View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
            view.setSystemUiVisibility(flags);
            return true;
        }
        return false;
    }

    // Dark status bar = light status bar theme for Android
    private boolean darkStatusBar() {
        if (supportsStatusBarTheming()) {
            final View view = getView();
            int flags = view.getSystemUiVisibility();
            flags |= View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
            view.setSystemUiVisibility(flags);
            return true;
        }
        return false;
    }

    // White status bar = default status bar theme for Android
    private boolean hasLightNavigationBar() {
        if (supportsNavigationBarTheming()) {
            final View view = getView();
            int flags = view.getSystemUiVisibility();
            return (flags & View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR) != View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR;
        }
        return false;
    }

    private boolean supportsNavigationBarTheming() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O;
    }

    private boolean lightNavigationBar() {
        if (supportsNavigationBarTheming()) {
            final View view = getView();
            int flags = view.getSystemUiVisibility();
            flags &= ~View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR;
            view.setSystemUiVisibility(flags);
            return true;
        }
        return false;
    }

    private boolean darkNavigationBar() {
        if (supportsNavigationBarTheming()) {
            final View view = getView();
            int flags = view.getSystemUiVisibility();
            flags |= View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR;
            view.setSystemUiVisibility(flags);
            return true;
        }
        return false;
    }

    private boolean supportsNavigationBarColorTheming() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;
    }

    private boolean changeNavigationBarColor(int red, int green, int blue) {
        if (supportsNavigationBarColorTheming()) {
            activity.getWindow().setNavigationBarColor(Color.argb(255, red, green, blue));
            return true;
        }
        return false;
    }

    private boolean hasANavigationBar() {
        final Resources res = activity.getResources();
        int id = res.getIdentifier("config_showNavigationBar", "bool", "android");
        return id > 0 && res.getBoolean(id);
    }

    private View getView() {
        return activity.getWindow().getDecorView();
    }

}


lib/screentheme.dart
import 'dart:async';
import 'dart:ui';

import 'package:flutter/services.dart';

class Platform {
  final _value;

  const Platform._internal(this._value);

  toString() => '$_value';

  static const Android = const Platform._internal('android');
  static const iOS = const Platform._internal('ios');
  static const both = const Platform._internal('both');
}

class ScreenTheme {
  static const MethodChannel _channel =
      const MethodChannel('g123k/screentheme');

  /// Will return true only if the status bar was changed
  /// On Android 6.0, the status bar items will become black
  static Future<bool> lightStatusBar({Platform platform = Platform.both}) async {
    bool res = await _channel.invokeMethod('changeStatusBarTheme',
        {'theme': 'light', 'platform': platform.toString()});
    return res;
  }

  /// Will return true only if the status bar was changed
  /// On Android 6.0, the status bar items will become white
  static Future<bool> darkStatusBar({Platform platform = Platform.both}) async {
    bool res = await _channel.invokeMethod('changeStatusBarTheme',
        {'theme': 'dark', 'platform': platform.toString()});
    return res;
  }

  static Future<bool> hasALightStatusBar({Platform platform = Platform.both}) async {
    bool res = await _channel
        .invokeMethod('hasLightStatusBar', {'platform': platform.toString()});
    return res;
  }

  static Future<bool> hasANavigationBar({Platform platform = Platform.both}) async {
    bool res = await _channel
        .invokeMethod('hasNavigationBar', {'platform': platform.toString()});
    return res;
  }

  /// Change the navigation bar color on Android only
  /// iOS: Will throw a PlatformException
  /// Android: Will only works on Android 5.0+
  static Future<bool> updateNavigationBarColor(Color color,
          {Platform platform = Platform.both}) async {
    bool res = await _channel.invokeMethod('changeNavigationBarColor', {
      'red': color.red,
      'blue': color.blue,
      'green': color.green,
      'platform': platform.toString()
    });
    return res;
  }

  /// Will return true only if the status bar was changed (only works on Android 8.0+)
  static Future<bool> lightNavigationBar({Platform platform = Platform.both}) async {
    bool res = await _channel.invokeMethod('changeNavigationBarTheme',
        {'theme': 'light', 'platform': platform.toString()});
    return res;
  }

  /// Will return true only if the status bar was changed (only works on Android 8.0+)
  static Future<bool> darkNavigationBar({Platform platform = Platform.both}) async {
    bool res = await _channel.invokeMethod('changeNavigationBarTheme',
        {'theme': 'dark', 'platform': platform.toString()});
    return res;
  }

  static Future<bool> hasALightNavigationBar(
      {Platform platform = Platform.both}) async {
    bool res = await _channel.invokeMethod(
        'hasLightNavigationBar', {'platform': platform.toString()});
    return res;
  }
}
